File: orchestrator/server/tracker.mjs
/* Task tracking with a declared backend (spec 083).
 *
 * The view reads and never writes: neither GitHub nor Linear offers concurrency control on an issue
 * write, so every write from here would be last-write-wins against whatever a teammate just did in
 * the web interface. Reading has no such failure mode.
 *
 * Every provider answers the same neutral row, and the vocabulary is deliberately not HTTP-shaped:
 * the local backend is a file read whose conflicts are git merges resolved by a person, and forcing
 * it to speak in status codes would distort the backend this project actually uses.
 */
import { readFile, stat } from 'node:fs/promises';
import path from 'node:path';
import { fail, resolveInRoot } from './store.mjs';
import { parseCredential, expiring, refresh } from './tracker-auth.mjs';
import { validateSchema } from './schema.mjs';

const PROBE_TTL_MS = 30000;          /* one poll every 30 s is about 5% of a Linear key's budget */
const CACHE_LIMIT = 64;
const cache = new Map();

/* A remote list is cached with in-flight coalescing, the shape the devices probe established: two
   callers arriving together share one request rather than spending the budget twice. The local
   backend is a file read and is never cached, because it is always current and a staleness
   indicator on something that cannot be stale would be a lie. */
async function coalesced(key, produce) {
  const hit = cache.get(key);
  if (hit && (hit.pending || Date.now() - hit.at < PROBE_TTL_MS)) return hit.value;
  const entry = { pending: true, at: Date.now(), value: produce() };
  cache.set(key, entry);
  if (cache.size > CACHE_LIMIT) cache.delete(cache.keys().next().value);
  try {
    entry.value = await entry.value;
    return entry.value;
  } catch (error) {
    cache.delete(key);
    throw error;
  } finally {
    entry.pending = false;
    entry.at = Date.now();
  }
}
export function forget() { cache.clear(); }

/* The token lives beside the workspace state and never in the committed declaration. Keyed by the
   declared project identity rather than a root id, so a person can create it by name. */
export async function credential(stateDirectory, project, options = {}) {
  if (!stateDirectory || !project) return null;
  const file = path.join(stateDirectory, 'trackers', `${project}.token`);
  let grant;
  try {
    grant = parseCredential(await readFile(file, 'utf8'));
  } catch (error) {
    if (error.code === 'ENOENT') return null;
    throw error;
  }
  if (!grant) return null;
  /* A signed-in grant is refreshed before it lapses; a pasted personal key never expires and is
     handed back untouched. A refresh that fails keeps the token it had, because Linear allows the
     request to be replayed for thirty minutes and a cleared grant could not use that. */
  if (expiring(grant)) grant = await refresh(stateDirectory, project, grant, options);
  return grant.accessToken ?? null;
}

/* The neutral row. State is (id, name, category) and never a boolean: a two-value enum cannot
   represent Linear's team-defined workflow states, and a local row's `passes` is never read back
   from a provider as truth. */
const CATEGORIES = ['backlog', 'unstarted', 'started', 'completed', 'canceled', 'blocked'];
const row = value => ({
  id: String(value.id),
  key: value.key ?? String(value.id),
  title: value.title ?? '',
  state: { id: value.stateId ?? '', name: value.stateName ?? '', category: value.category ?? 'backlog' },
  priority: value.priority ?? null,
  labels: value.labels ?? [],
  assignee: value.assignee ?? null,
  url: value.url ?? null,
  updatedAt: value.updatedAt ?? null,
  blockedBy: value.blockedBy ?? [],
  /* Added for spec 103: the prompt a spawned agent is seeded with says what "done" means, and only
     the local backend has that written down. A remote row answers with an empty list rather than
     with the issue body, which is prose rather than criteria. */
  criteria: value.criteria ?? [],
  /* Added for spec 116: what a criterion claims and what proves it are one question, and reading
     them apart is what left the Tasks tab unable to answer "which test backs this". Same empty-list
     rule as criteria, and for the same reason — an issue body is prose, not an evidence list. */
  evidence: value.evidence ?? [],
  /* Added for spec 117: the entries a project's own tests manifest records for this task. rEngine
     joins them by the provider's key and runs nothing; an entry never moves a row. */
  tests: value.tests ?? [],
});

/* Local. Readiness follows the same rule tools/features.py applies, so the view and the command
   line cannot disagree about what is blocked. */
function localState(feature, byId) {
  if (feature.passes) return { stateId: 'passing', stateName: 'passing', category: 'completed' };
  const unmet = (feature.dependencies ?? []).some(id => !byId.get(id)?.passes);
  if (unmet) return { stateId: 'blocked', stateName: 'blocked', category: 'blocked' };
  return { stateId: 'ready', stateName: 'ready', category: 'unstarted' };
}
async function localRows(root, block) {
  const name = block.inventory ?? 'features.json';
  let parsed;
  try {
    parsed = JSON.parse(await readFile(path.join(root.path, name), 'utf8'));
  } catch (error) {
    if (error.code === 'ENOENT') return { rows: [], unavailable: `${name} is not in this project.` };
    return { rows: [], invalid: [`${name}: ${error.message}`] };
  }
  const features = Array.isArray(parsed?.features) ? parsed.features : [];
  const byId = new Map(features.map(feature => [feature.id, feature]));
  return {
    rows: features.map(feature => row({
      id: feature.id,
      key: `F${feature.id}`,
      title: feature.description,
      ...localState(feature, byId),
      priority: feature.priority ?? null,
      labels: [feature.milestone, feature.category].filter(Boolean),
      assignee: feature.owner_workspace ?? null,
      blockedBy: (feature.dependencies ?? []).map(id => `F${id}`),
      criteria: Array.isArray(feature.acceptance_criteria) ? feature.acceptance_criteria : [],
      evidence: Array.isArray(feature.evidence) ? feature.evidence.filter(item => typeof item === 'string') : [],
    })),
  };
}

/* Linear. A personal API key sends the token bare, without a Bearer prefix, which is the one thing
   about its auth that surprises everyone. States are objects carrying a category, so they reach the
   neutral row without being flattened into open and closed. */
/* The whole filter is one variable, so this document is a fixed string whatever the declaration
   narrows to; nothing is interpolated into it. See linearFilter for what goes in. */
const LINEAR_QUERY = `query Issues($filter: IssueFilter, $first: Int!) {
  issues(first: $first, filter: $filter, orderBy: updatedAt) {
    nodes {
      id identifier title url priority updatedAt
      state { id name type }
      assignee { displayName }
      labels(first: 10) { nodes { name } }
      relations(first: 20) { nodes { type relatedIssue { identifier } } }
    }
  }
}`;
const LINEAR_PRIORITY = [null, 'urgent', 'high', 'medium', 'low'];

/* An undeclared narrowing contributes no clause, so the two-clause object below is byte-identical to
   what the previous document produced and an existing declaration asks exactly what it always did.
   The project clause stays null-safe: present and null leaves the team unfiltered (spec 100). */
function linearFilter(block) {
  const filter = { team: { key: { eq: block.team } }, project: { name: { eq: block.project ?? null } } };
  /* "me" is the one assignee value that is not a name: it asks Linear who the token belongs to, so a
     personal declaration keeps working when someone else's token reads it. */
  if (block.assignee !== undefined) filter.assignee = block.assignee === 'me' ? { isMe: { eq: true } } : { displayName: { eq: block.assignee } };
  /* Declared categories are the row's own vocabulary, which is Linear's state type. */
  if (block.states !== undefined) filter.state = { type: { in: block.states } };
  return filter;
}

async function linearRows(block, token, fetchImpl) {
  if (!token) return { rows: [], denied: 'Not signed in to Linear.', signIn: 'linear' };
  const response = await fetchImpl('https://api.linear.app/graphql', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: token },
    body: JSON.stringify({ query: LINEAR_QUERY, variables: { filter: linearFilter(block), first: 100 } }),
  });
  if (response.status === 401 || response.status === 403) return { rows: [], denied: 'Linear refused the token.' };
  if (!response.ok) return { rows: [], unavailable: `Linear answered ${response.status}.` };
  const body = await response.json();
  if (body.errors?.length) {
    const message = body.errors.map(error => error.message).join('; ');
    /* Linear reports its rate limit as a 400 carrying a RATELIMITED error rather than a 429. */
    if (/ratelimit/i.test(message)) return { rows: [], unavailable: 'Linear rate limit reached; the list refreshes shortly.' };
    return { rows: [], invalid: body.errors.map(error => error.message) };
  }
  const nodes = body.data?.issues?.nodes ?? [];
  return {
    rows: nodes.map(issue => row({
      id: issue.id,
      key: issue.identifier,
      title: issue.title,
      url: issue.url,
      stateId: issue.state?.id,
      stateName: issue.state?.name,
      category: issue.state?.type,
      priority: LINEAR_PRIORITY[issue.priority] ?? null,
      labels: (issue.labels?.nodes ?? []).map(label => label.name),
      assignee: issue.assignee?.displayName ?? null,
      updatedAt: issue.updatedAt,
      blockedBy: (issue.relations?.nodes ?? [])
        .filter(relation => relation.type === 'blocks')
        .map(relation => relation.relatedIssue?.identifier)
        .filter(Boolean),
    })),
  };
}

/* GitHub. State is open or closed with a reason, so the category is derived rather than read. */
async function githubRows(block, token, fetchImpl) {
  if (!token) return { rows: [], denied: `No GitHub token. Put one in the workspace state directory as trackers/${block.identity}.token` };
  const url = `https://api.github.com/repos/${block.repository}/issues?state=all&sort=updated&direction=desc&per_page=100`;
  const response = await fetchImpl(url, {
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
    },
  });
  if (response.status === 401 || response.status === 403) return { rows: [], denied: 'GitHub refused the token.' };
  if (response.status === 404) return { rows: [], invalid: [`${block.repository} is not reachable with this token.`] };
  if (!response.ok) return { rows: [], unavailable: `GitHub answered ${response.status}.` };
  const issues = (await response.json()).filter(issue => !issue.pull_request);
  return {
    rows: issues.map(issue => row({
      id: issue.id,
      key: `#${issue.number}`,
      title: issue.title,
      url: issue.html_url,
      stateId: issue.state,
      stateName: issue.state_reason ?? issue.state,
      category: issue.state === 'closed' ? (issue.state_reason === 'not_planned' ? 'canceled' : 'completed') : 'unstarted',
      labels: (issue.labels ?? []).map(label => (typeof label === 'string' ? label : label.name)),
      assignee: issue.assignee?.login ?? null,
      updatedAt: issue.updated_at,
    })),
  };
}

/* The project's tests manifest (contract 10, spec 117). Read here rather than in the declaration
   because it is the project's own artifact with its own lifetime, exactly as the local inventory is.
   Nothing in here opens a test, runs anything, or lets an entry change a row's state. */
const TESTS_SCHEMA = JSON.parse(await readFile(new URL('../../contracts/task-tests-v1.schema.json', import.meta.url), 'utf8'));

/* The checkout's own revision, read from files and never by running git: a subprocess for a display
   detail is a cost and a permission this reader should not take. A layout it cannot follow answers
   null, so "stale" and "cannot tell" stay different answers. */
async function headCommit(rootPath) {
  try {
    const head = (await readFile(path.join(rootPath, '.git', 'HEAD'), 'utf8')).trim();
    if (/^[0-9a-f]{40}$/.test(head)) return head;
    const ref = head.startsWith('ref: ') ? head.slice(5).trim() : null;
    if (!ref) return null;
    const value = (await readFile(path.join(rootPath, '.git', ref), 'utf8')).trim();
    return /^[0-9a-f]{40}$/.test(value) ? value : null; /* packed refs are a follow-up; unknown beats a guess */
  } catch { return null; }
}

async function withTests(root, declared, result) {
  const manifest = declared?.tests?.manifest;
  if (!manifest) return result;
  let parsed;
  try {
    parsed = JSON.parse(await readFile(path.join(root.path, manifest), 'utf8'));
  } catch (error) {
    const missing = error.code === 'ENOENT';
    return { ...result, testsError: `${manifest}: ${missing ? 'the declared tests manifest is not in this project.' : error.message}` };
  }
  const problems = validateSchema(TESTS_SCHEMA, parsed, TESTS_SCHEMA, '$');
  if (problems.length) return { ...result, testsError: `${manifest}: ${problems.slice(0, 3).join('; ')}` };

  const byTask = new Map();
  for (const entry of parsed.entries ?? []) {
    if (!byTask.has(entry.task)) byTask.set(entry.task, []);
    byTask.get(entry.task).push(entry);
  }
  /* A manifest drifts from its inventory the moment a criterion is renumbered, and a claim pointing
     past the task's criteria reads as coverage it does not have — worse than claiming none. Only a
     provider that knows its own criteria can be checked, so the others carry the index unjudged. */
  const drift = [];
  for (const row of result.rows ?? []) {
    for (const entry of byTask.get(row.key) ?? []) {
      for (const index of entry.criteria ?? []) {
        if (row.criteria.length && index > row.criteria.length) {
          drift.push(`${manifest}: ${row.key} claims criterion ${index}, but the task has ${row.criteria.length} criterion${row.criteria.length === 1 ? '' : 's'}`);
        }
      }
    }
  }
  const head = await headCommit(root.path);
  /* An artifact is answered for HERE rather than when someone clicks it, so a row can say "missing"
     or "outside this project" instead of a click failing. rEngine still opens nothing it was not
     asked to open and still produces nothing: this is a resolve and a stat (spec 126). */
  const outside = [];
  const settle = async artifact => {
    const label = typeof artifact.label === 'string' ? artifact.label : '';
    try {
      const { relative } = await resolveInRoot(root, artifact.path, true);
      let state = 'ok';
      try { await stat(path.join(root.path, relative)); } catch { state = 'missing'; }
      return { path: relative, label, state };
    } catch {
      outside.push(artifact.path);
      return { path: artifact.path, label, state: 'outside' };
    }
  };
  const rows = await Promise.all((result.rows ?? []).map(async row => ({
    ...row,
    tests: await Promise.all((byTask.get(row.key) ?? []).map(async entry => ({
      ...entry,
      /* The field the format exists for: a green run says a command went green, and only a sabotage
         row says the test can go red for its own reason (AGENTS.md). Never collapsed into one word. */
      proven: Array.isArray(entry.sabotage) && entry.sabotage.length > 0,
      ...(Array.isArray(entry.last?.artifacts)
        ? { last: { ...entry.last, artifacts: await Promise.all(entry.last.artifacts.map(settle)) } }
        : {}),
    }))),
  })));
  if (outside.length) drift.push(`${manifest}: ${outside.slice(0, 3).map(p => `${p} is outside this project`).join('; ')}`);
  return {
    ...result, rows,
    tests: { at: parsed.at ?? null, commit: parsed.commit ?? null, count: parsed.entries?.length ?? 0,
             current: head && parsed.commit ? head === parsed.commit : null },
    ...(drift.length ? { testsError: drift.slice(0, 3).join('; ') } : {}),
  };
}

export async function projectTracker(root, declared, options = {}) {
  const fetchImpl = options.fetch ?? globalThis.fetch;
  const base = { rootId: root.id, declared: declared.declared === true, provider: null, rows: [], categories: CATEGORIES };
  /* A project that declares nothing at all still has its own inventory if it keeps one, which is
     the default backend and the one this repository uses. */
  if (!declared.declared) return { ...base, provider: 'local', ...(await localRows(root, {})), fresh: true };
  if (declared.error) return { ...base, error: declared.error };
  if (declared.trackerError) return { ...base, contract: declared.contract, error: declared.trackerError };
  /* A project that declares no tracker still has its own inventory, which is the default backend. */
  const block = declared.tracker ?? { provider: 'local' };
  /* `identity` is the declared project name the token file is keyed by; `project` inside the block
     is Linear's project filter. Naming both `project` made the filter silently take the token's
     value, which the tests caught by asserting the variable that reaches the query. */
  const named = { ...block, identity: declared.project ?? root.id };
  const result = { ...base, contract: declared.contract, provider: block.provider };
  if (block.provider === 'local') return withTests(root, declared, { ...result, ...(await localRows(root, block)), fresh: true });

  if (typeof fetchImpl !== 'function') fail('This build cannot reach a network tracker.', 501);
  const token = await credential(options.stateDirectory, named.identity, options);
  /* The narrowing is part of the key: two declarations that ask different questions are different
     questions, and answering the second from the first's entry would be wrong rather than stale. */
  const key = [root.id, block.provider, block.repository ?? block.team, block.project ?? '', block.assignee ?? '', (block.states ?? []).join(',')].join(' ');
  if (options.refresh) cache.delete(key);
  const produce = () => (block.provider === 'linear'
    ? linearRows(named, token, fetchImpl)
    : githubRows(named, token, fetchImpl));
  const entry = await coalesced(key, produce);
  const at = cache.get(key)?.at ?? Date.now();
  return withTests(root, declared, { ...result, ...entry, fresh: Date.now() - at < PROBE_TTL_MS, checkedAt: new Date(at).toISOString() });
}

File: red/red-project/src/tracker.rs
//! A project's own task inventory, and the tests it says are behind each task (F153; specs 100,
//! 103, 116, 117).
//!
//! Only the LOCAL backend is here. The remote providers need a network client, and F154 owns that
//! decision — so what this holds is the backend this repository itself uses: `features.json` read as
//! neutral rows, with the readiness `tools/features.py` applies, so the Tasks tab and the command
//! line cannot disagree about what is blocked.
//!
//! Joined onto it is the tests manifest of spec 117: what a criterion claims and what proves it.
//! rEngine READS it and runs nothing — no entry moves a row, and a manifest saying everything failed
//! leaves every task where its provider put it. The one field the format exists for is `proven`: a
//! green run says a command went green, and only a sabotage row says the test can go red for its
//! own reason (AGENTS.md). The two are never collapsed into one word.

use std::path::Path;

use serde_json::{json, Value};

use crate::rules::object;

/// State is (id, name, category) and never a boolean: a two-value enum cannot represent Linear's
/// team-defined workflow states, and a local row's `passes` is never read back as truth.
pub const CATEGORIES: [&str; 6] = ["backlog", "unstarted", "started", "completed", "canceled", "blocked"];

const TESTS_SCHEMA: &str = include_str!("../../../contracts/task-tests-v1.schema.json");

fn text<'a>(value: &'a Value, key: &str) -> &'a str {
    value.get(key).and_then(Value::as_str).unwrap_or_default()
}

fn strings(value: &Value, key: &str) -> Vec<String> {
    value
        .get(key)
        .and_then(Value::as_array)
        .map(|items| items.iter().filter_map(Value::as_str).map(str::to_string).collect())
        .unwrap_or_default()
}

/// The neutral row every provider is flattened into.
/// The same neutral row, for a provider's issue. Public because `tracker_remote` builds one and the
/// two must be the same shape: a GitHub row and a local one are read by the same pane.
pub fn remote_row(id: String, key: String, title: Value, state: (&str, &str, &str), fields: Vec<(&'static str, Value)>) -> Value {
    row(id, key, title, state, fields)
}

fn row(id: String, key: String, title: Value, state: (&str, &str, &str), fields: Vec<(&'static str, Value)>) -> Value {
    let mut out = vec![
        ("id", json!(id)),
        ("key", json!(key)),
        ("title", title),
        ("state", json!({ "id": state.0, "name": state.1, "category": state.2 })),
        ("priority", Value::Null),
        ("labels", json!([])),
        ("assignee", Value::Null),
        ("url", Value::Null),
        ("updatedAt", Value::Null),
        ("blockedBy", json!([])),
        ("criteria", json!([])),
        ("evidence", json!([])),
        ("tests", json!([])),
    ];
    for (name, value) in fields {
        if let Some(slot) = out.iter_mut().find(|(key, _)| *key == name) {
            slot.1 = value;
        }
    }
    object(out)
}

/// Readiness follows the rule `tools/features.py` applies, so the view and the command line cannot
/// disagree about what is blocked.
fn local_state(feature: &Value, features: &[Value]) -> (&'static str, &'static str, &'static str) {
    if feature.get("passes").and_then(Value::as_bool) == Some(true) {
        return ("passing", "passing", "completed");
    }
    let unmet = feature
        .get("dependencies")
        .and_then(Value::as_array)
        .map(|ids| {
            ids.iter().any(|id| {
                features
                    .iter()
                    .find(|feature| feature.get("id") == Some(id))
                    .and_then(|feature| feature.get("passes").and_then(Value::as_bool))
                    != Some(true)
            })
        })
        .unwrap_or(false);
    if unmet {
        ("blocked", "blocked", "blocked")
    } else {
        ("ready", "ready", "unstarted")
    }
}

/// The project's own inventory, or the named reason there is none to read.
pub fn local_rows(root_path: &str, block: &Value) -> Value {
    let name = block.get("inventory").and_then(Value::as_str).unwrap_or("features.json");
    let file = Path::new(root_path).join(name);
    let text = match std::fs::read_to_string(&file) {
        Ok(text) => text,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
            return json!({ "rows": [], "unavailable": format!("{name} is not in this project.") });
        }
        Err(error) => return json!({ "rows": [], "invalid": [format!("{name}: {error}")] }),
    };
    let parsed: Value = match serde_json::from_str(&text) {
        Ok(value) => value,
        Err(error) => return json!({ "rows": [], "invalid": [format!("{name}: {error}")] }),
    };
    let features: Vec<Value> = parsed.get("features").and_then(Value::as_array).cloned().unwrap_or_default();
    let rows: Vec<Value> = features
        .iter()
        .map(|feature| {
            let id = feature.get("id").map(js_id).unwrap_or_default();
            let dependencies: Vec<String> = feature
                .get("dependencies")
                .and_then(Value::as_array)
                .map(|ids| ids.iter().map(|id| format!("F{}", js_id(id))).collect())
                .unwrap_or_default();
            row(
                id.clone(),
                format!("F{id}"),
                feature.get("description").cloned().unwrap_or_else(|| json!("")),
                local_state(feature, &features),
                vec![
                    ("priority", feature.get("priority").cloned().unwrap_or(Value::Null)),
                    (
                        "labels",
                        json!([feature.get("milestone"), feature.get("category")]
                            .into_iter()
                            .flatten()
                            .filter(|value| !value.is_null() && value.as_str() != Some(""))
                            .cloned()
                            .collect::<Vec<_>>()),
                    ),
                    ("assignee", feature.get("owner_workspace").cloned().unwrap_or(Value::Null)),
                    ("blockedBy", json!(dependencies)),
                    ("criteria", feature.get("acceptance_criteria").filter(|value| value.is_array()).cloned().unwrap_or_else(|| json!([]))),
                    (
                        "evidence",
                        json!(feature
                            .get("evidence")
                            .and_then(Value::as_array)
                            .map(|items| items.iter().filter(|item| item.is_string()).cloned().collect::<Vec<_>>())
                            .unwrap_or_default()),
                    ),
                ],
            )
        })
        .collect();
    json!({ "rows": rows })
}

/// `String(value)` for an inventory id, which may be a number or a string.
fn js_id(value: &Value) -> String {
    match value {
        Value::String(text) => text.clone(),
        other => other.to_string(),
    }
}

/// The revision this checkout is at, so a reader can say whether a manifest describes it.
fn head_commit(root_path: &str) -> Option<String> {
    let git = Path::new(root_path).join(".git");
    let head = std::fs::read_to_string(git.join("HEAD")).ok()?.trim().to_string();
    if is_sha(&head) {
        return Some(head);
    }
    let reference = head.strip_prefix("ref: ")?.trim().to_string();
    /* Packed refs are a follow-up; unknown beats a guess. */
    let value = std::fs::read_to_string(git.join(reference)).ok()?.trim().to_string();
    is_sha(&value).then_some(value)
}

fn is_sha(value: &str) -> bool {
    value.len() == 40 && value.bytes().all(|byte| byte.is_ascii_hexdigit() && !byte.is_ascii_uppercase())
}

/// The manifest a declaration names, joined onto the rows it keys.
pub fn with_tests(root_path: &str, declared: &Value, result: &Value) -> Value {
    let manifest = declared.get("tests").and_then(|tests| tests.get("manifest")).and_then(Value::as_str);
    let Some(manifest) = manifest.filter(|name| !name.is_empty()) else { return result.clone() };
    let file = Path::new(root_path).join(manifest);
    let parsed: Value = match std::fs::read_to_string(&file) {
        Ok(text) => match serde_json::from_str(&text) {
            Ok(value) => value,
            Err(error) => return with_error(result, format!("{manifest}: {error}")),
        },
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
            return with_error(result, format!("{manifest}: the declared tests manifest is not in this project."));
        }
        Err(error) => return with_error(result, format!("{manifest}: {error}")),
    };
    let schema: Value = serde_json::from_str(TESTS_SCHEMA).expect("the tests contract");
    let problems = red_store::schema::validate_schema(&schema, &parsed);
    if !problems.is_empty() {
        return with_error(result, format!("{manifest}: {}", problems.iter().take(3).cloned().collect::<Vec<_>>().join("; ")));
    }

    let entries: Vec<Value> = parsed.get("entries").and_then(Value::as_array).cloned().unwrap_or_default();
    let for_task = |key: &str| -> Vec<Value> { entries.iter().filter(|entry| text(entry, "task") == key).cloned().collect() };
    let rows: Vec<Value> = result.get("rows").and_then(Value::as_array).cloned().unwrap_or_default();

    /* A manifest drifts from its inventory the moment a criterion is renumbered, and a claim
       pointing past the task's criteria reads as coverage it does not have — worse than claiming
       none. Only a provider that knows its own criteria can be checked, so the others carry the
       index unjudged. */
    let mut drift: Vec<String> = Vec::new();
    for row in &rows {
        let criteria = row.get("criteria").and_then(Value::as_array).map(Vec::len).unwrap_or(0);
        for entry in for_task(text(row, "key")) {
            for index in entry.get("criteria").and_then(Value::as_array).map(Vec::as_slice).unwrap_or_default() {
                let index = index.as_u64().unwrap_or(0) as usize;
                if criteria > 0 && index > criteria {
                    drift.push(format!(
                        "{manifest}: {} claims criterion {index}, but the task has {criteria} criterion{}",
                        text(row, "key"),
                        if criteria == 1 { "" } else { "s" }
                    ));
                }
            }
        }
    }

    /* An artifact is answered for HERE rather than when someone clicks it, so a row can say
       "missing" or "outside this project" instead of a click failing. rEngine still opens nothing
       it was not asked to open and still produces nothing: this is a resolve and a stat (spec 126). */
    let mut outside: Vec<String> = Vec::new();
    let mut joined: Vec<Value> = Vec::new();
    for row in &rows {
        let mut out = row.as_object().cloned().unwrap_or_default();
        let tests: Vec<Value> = for_task(text(row, "key"))
            .into_iter()
            .map(|entry| {
                let mut entry = entry.as_object().cloned().unwrap_or_default();
                let proven = entry.get("sabotage").and_then(Value::as_array).is_some_and(|rows| !rows.is_empty());
                if let Some(artifacts) = entry.get("last").and_then(|last| last.get("artifacts")).and_then(Value::as_array).cloned() {
                    let settled: Vec<Value> = artifacts.iter().map(|artifact| settle(root_path, artifact, &mut outside)).collect();
                    let mut last = entry.get("last").and_then(Value::as_object).cloned().unwrap_or_default();
                    last.insert("artifacts".into(), json!(settled));
                    entry.insert("proven".into(), json!(proven));
                    entry.insert("last".into(), Value::Object(last));
                } else {
                    entry.insert("proven".into(), json!(proven));
                }
                Value::Object(entry)
            })
            .collect();
        out.insert("tests".into(), json!(tests));
        joined.push(Value::Object(out));
    }
    if !outside.is_empty() {
        drift.push(format!(
            "{manifest}: {}",
            outside.iter().take(3).map(|path| format!("{path} is outside this project")).collect::<Vec<_>>().join("; ")
        ));
    }
    let head = head_commit(root_path);
    let commit = parsed.get("commit").and_then(Value::as_str);
    let mut out = result.as_object().cloned().unwrap_or_default();
    out.insert("rows".into(), json!(joined));
    out.insert(
        "tests".into(),
        json!({
            "at": parsed.get("at").cloned().unwrap_or(Value::Null),
            "commit": parsed.get("commit").cloned().unwrap_or(Value::Null),
            "count": entries.len(),
            "current": match (head.as_deref(), commit) {
                (Some(head), Some(commit)) => json!(head == commit),
                _ => Value::Null,
            },
        }),
    );
    if !drift.is_empty() {
        out.insert("testsError".into(), json!(drift.iter().take(3).cloned().collect::<Vec<_>>().join("; ")));
    }
    Value::Object(out)
}

fn settle(root_path: &str, artifact: &Value, outside: &mut Vec<String>) -> Value {
    let label = artifact.get("label").and_then(Value::as_str).unwrap_or("");
    let declared = text(artifact, "path");
    match red_store::store::resolve_in_root(root_path, declared, true) {
        Ok((_, relative)) => {
            let state = if Path::new(root_path).join(&relative).exists() { "ok" } else { "missing" };
            json!({ "path": relative, "label": label, "state": state })
        }
        Err(_) => {
            outside.push(declared.to_string());
            json!({ "path": declared, "label": label, "state": "outside" })
        }
    }
}

fn with_error(result: &Value, message: String) -> Value {
    let mut out = result.as_object().cloned().unwrap_or_default();
    out.insert("testsError".into(), json!(message));
    Value::Object(out)
}

/// The whole local answer, in the shape the route gives it.
/// The same answer, for a project whose tracker is somebody else's server (F154, spec 083).
///
/// `state_directory` is where the credential lives — beside the WORKSPACE state and never in the
/// committed declaration — and `identity` is the declared project NAME the token file is keyed by,
/// so a person can create it by name and a checkout that moved keeps its tracker.
///
/// **`identity` and the block's own `project` are different things.** The block's `project` is
/// Linear's project filter. Naming both `project` made the filter silently take the token's value.
pub fn remote_tracker(
    root_id: &str,
    root_path: &str,
    declared: &Value,
    state_directory: &str,
    credential: Option<&str>,
    fetching: &dyn crate::tracker_remote::Fetching,
    now_ms: i64,
) -> Value {
    let _ = state_directory;
    let block = declared.get("tracker").cloned().unwrap_or_else(|| json!({ "provider": "local" }));
    let provider = block.get("provider").and_then(Value::as_str).unwrap_or("local").to_string();
    let mut named = block.as_object().cloned().unwrap_or_default();
    named.insert(
        "identity".to_string(),
        declared.get("project").cloned().filter(|name| !name.is_null()).unwrap_or_else(|| json!(root_id)),
    );
    let named = Value::Object(named);
    let answered = match provider.as_str() {
        "linear" => crate::tracker_remote::linear_rows(&named, credential, fetching),
        _ => crate::tracker_remote::github_rows(&named, credential, fetching),
    };
    let base = object(vec![
        ("rootId", json!(root_id)),
        ("declared", json!(true)),
        ("provider", json!(provider)),
        ("rows", json!([])),
        ("categories", json!(CATEGORIES)),
        ("contract", declared.get("contract").cloned().unwrap_or(Value::Null)),
    ]);
    let joined = merge(base, &answered, vec![
        ("fresh", json!(true)),
        ("checkedAt", json!(red_core::time::iso(now_ms))),
    ]);
    /* The manifest is joined onto a remote provider's rows by the same reader, so a GitHub row and a
       local one carry their evidence in the same shape. */
    with_tests(root_path, declared, &joined)
}

pub fn project_tracker(root_id: &str, root_path: &str, declared: &Value) -> Value {
    let base = |provider: Value, extra: Vec<(&'static str, Value)>| {
        let mut out = vec![
            ("rootId", json!(root_id)),
            ("declared", json!(declared.get("declared").and_then(Value::as_bool) == Some(true))),
            ("provider", provider),
            ("rows", json!([])),
            ("categories", json!(CATEGORIES)),
        ];
        out.extend(extra);
        object(out)
    };
    /* A project that declares nothing at all still has its own inventory if it keeps one, which is
       the default backend and the one this repository uses. */
    if declared.get("declared").and_then(Value::as_bool) != Some(true) {
        return merge(base(json!("local"), vec![]), &local_rows(root_path, &json!({})), vec![("fresh", json!(true))]);
    }
    if let Some(said) = declared.get("error").and_then(Value::as_str) {
        return base(Value::Null, vec![("error", json!(said))]);
    }
    if let Some(said) = declared.get("trackerError").and_then(Value::as_str) {
        return base(Value::Null, vec![("contract", declared.get("contract").cloned().unwrap_or(Value::Null)), ("error", json!(said))]);
    }
    let block = declared.get("tracker").cloned().unwrap_or_else(|| json!({ "provider": "local" }));
    let result = base(
        block.get("provider").cloned().unwrap_or_else(|| json!("local")),
        vec![("contract", declared.get("contract").cloned().unwrap_or(Value::Null))],
    );
    let local = merge(result, &local_rows(root_path, &block), vec![("fresh", json!(true))]);
    with_tests(root_path, declared, &local)
}

/// `{ ...result, ...rows, ...extra }` with the key order a spread gives.
fn merge(result: Value, rows: &Value, extra: Vec<(&'static str, Value)>) -> Value {
    let mut out = result.as_object().cloned().unwrap_or_default();
    if let Some(map) = rows.as_object() {
        for (key, value) in map {
            out.insert(key.clone(), value.clone());
        }
    }
    for (key, value) in extra {
        out.insert(key.into(), value);
    }
    Value::Object(out)
}
